ssl/ssl_lib.c

```diff
@@ -2205,20 +2212,41 @@ long SSL_CTX_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
     case SSL_CTRL_SET_SESS_CACHE_SIZE:
         if (larg < 0)
             return 0;
-        l = (long)ctx->session_cache_size;
-        ctx->session_cache_size = (size_t)larg;
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = (long)ctx->session_cache->size;
+        ctx->session_cache->size = (size_t)larg;
+        CRYPTO_THREAD_unlock(ctx->session_cache->lock);
+        CRYPTO_THREAD_unlock(ctx->lock);
         return l;
     case SSL_CTRL_GET_SESS_CACHE_SIZE:
-        return (long)(ctx->session_cache_size);
+        CRYPTO_THREAD_read_lock(ctx->lock);
+        l = (long)ctx->session_cache->size;
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return l;
+
     case SSL_CTRL_SET_SESS_CACHE_MODE:
-        l = ctx->session_cache_mode;
-        ctx->session_cache_mode = larg;
-        return (l);
+        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(ctx->session_cache->lock);
+        l = ctx->session_cache->mode;
+        ctx->session_cache->mode = larg;
```

Comment:
That makes some sense, but it's difficult to control setting those parameters through the SSL_CTX. The original SSL_CTX doesn't know that it was shared with any other SSL_CTX (or not); using `reference > 1` is not reliable depending on how the cache is created.

I see two ways this mechanism could be initialized:

```
SSL_SESSION_CACHE *cache = SSL_SESSION_CACHE_new();
SSL_CTX *ctx1 = SSL_CTX_new_ex(method, cache);
SSL_CTX *ctx2 = SSL_CTX_new_ex(method, cache);
...
SSL_SESSION_CACHE_free(cache, NULL);
```

or

```
SSL_CTX *ctx1 = SSL_CTX_new(method);
SSL_SESSION_CACHE *cache = SSL_CTX_get1_session_cache(ctx);
SSL_CTX *ctx2 = SSL_CTX_new_ex(method, cache);
SSL_SESSION_CACHE_free(cache, NULL);
```